x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/SqlVersionUtils.java

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.sql.jdbc;
+
+import org.elasticsearch.Build;
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.xpack.sql.proto.SqlVersion;
+
+public final class SqlVersionUtils {
+
+    public static final SqlVersion CURRENT = SqlVersion.fromString(Build.current().version());
+
+    // Translating a TransportVersion to a SqlVersion/Version must go through the former's string representation, which involves a
+    // mapping; the .id() can't be used directly.
+    public static SqlVersion from(TransportVersion transportVersion) {
+        return SqlVersion.fromTransportString(transportVersion.toReleaseVersion());
```

Comment:
IIUC, this function in particular is "safe", as `toReleaseVersion()` will always produce a string that can be translated to a `SqlVersion`, i.e. also in serverless -- but let me know if this assumption is wrong please.
What I think will fail is `Build.current().version()` for a "CURRENT" value, since that will only be a hash in Serverless. Thanks for pointing that out.

Reply:
This is not necessarily the case - `toReleaseVersion` can produce ranges of versions. It's only meant for text strings, not for parsing back into semantic versions.

Reply:
> `toReleaseVersion` can produce ranges of versions.

That's catered for already, and my interpretation of the range is that a version has been added in between those two stack release versions. And with that, SqlVersion will considerer the upper bound only, since I believe a stack release (of an SQL client) would embed - and later report to the server - that version.
But let me know if this is wrong, please.

Reply:
Even so, `toReleaseVersion` is not intended for parsing back into semantic versions (output can also be things like `8.5.0-snapshot[512000]`) - this use is extremely fragile.

Reply:
> Even so, toReleaseVersion is not intended for parsing back into semantic versions (output can also be things like 8.5.0-snapshot[512000]) - this use is extremely fragile.

Thanks Simon. That's also been catered for ([fwiw](https://github.com/elastic/elasticsearch/pull/112094/files#diff-992e748db302adb873c258151a11f649d5119d753536a04f52e1555f22e1e0a6R21)).

I'm trying another approach, introducing a server-side SQL version, id-based, which will hopefully address these concerns ..in server. But making the SemVer disappear entirely isn't achievable IMO, since the clients have been (and still are) released under that scheme. 
SQL API doesn't make use of the [REST API compatibility](https://www.elastic.co/guide/en/elasticsearch/reference/current/rest-api-compatibility.html) scheme (as it required finer control than that offers) and has its own [backwards compatibility guarantees](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-jdbc.html#jdbc-compatibility) (which will likely need to update; at least in serverless).

Reply:
This has been redone. The Build's version is still considered, but if not available, a `SERVER_COMPAT_VERSION`, which is the latest known SqlVersion (fixed at `8.16.0`) will be made use of -- the required server compatibility version.